source/common/network/connection_impl.cc

```diff
@@ -123,35 +133,33 @@ void ConnectionImpl::close(ConnectionCloseType type) {
     //          ConnectionManagerImpl::checkForDeferredClose()
     //     2) A second close is issued by a subsequent call to
     //        ConnectionManagerImpl::checkForDeferredClose() prior to returning from onData()
-    if (delayed_close_) {
+    if (inDelayedClose()) {
+      // Validate that a delayed close timer is already enabled unless it was disabled via
+      // configuration.
+      ASSERT(!delayed_close_timeout_set || delayed_close_timer_ != nullptr);
+      // Validate that the same close type is used when multiple close()s are issued. An edge case
```

Comment:
Do we make it clear in the class definition what transitions are allowed?

Reply:
Done. Added a comment to the DelayedCloseState enum declaration.